### DISCUSSION TITLE
Infer ComponentProps of Dialog

### DISCUSSION THREAD

**Discussion by giovanniincammicia** [tailwindlabs/headlessui#601] [Help]
Hi, I'm trying to create a custom Dialog using headlessui Dialog. My custom Dialog should accept the same props as headlessui one. Using typescript, I'm trying to infer the Props type as I usually do, but It's not working as expected.

```
import { Dialog as UIDialog } from '@headlessui/react';

export default function Dialog ({ open, onClose }: React.ComponentProps<typeof UIDialog>) { ... }
```

This is the error I'm getting:
```
Type '(<TTag extends ElementType<any> = "div">(props: (Pick<PropsOf<TTag>, Exclude<keyof PropsOf<TTag>, "children" | "id" | "as" | "refName" | "className">> & ... 4 more ... & { ...; }) | ... 10 more ... | (Pick<...> & ... 4 more ... & { ...; }), ref: Ref<...>) => Element) & { ...; } & { ...; }' does not satisfy the constraint 'keyof IntrinsicElements | JSXElementConstructor<any>'.
  Type '(<TTag extends ElementType<any> = "div">(props: (Pick<PropsOf<TTag>, Exclude<keyof PropsOf<TTag>, "children" | "id" | "as" | "refName" | "className">> & ... 4 more ... & { ...; }) | ... 10 more ... | (Pick<...> & ... 4 more ... & { ...; }), ref: Ref<...>) => Element) & { ...; } & { ...; }' is not assignable to type '(props: any) => ReactElement<any, any> | null'.
```
  
Any idea? Is this a bug or am I doing something wrong? Thanks.

**Comment by giovanniincammicia** [comment]
Anyone?

**Comment by jrozbicki** [comment]
I'm also looking to create custom component from Dialog and extend its props, haven't found a way to do this properly unfortunately. Would be nice to be able to have `DialogProps` and be able to extend custom interface with that

**Reply by baronnoraz** [reply]
I second this.  I was fully expecting to have access to something like `DialogProps`.  Without it, I've been struggling to make my custom component wrap the Dialog component and make typescript happy.

**Reply by MerlinMason** [reply]
I came here looking to do exactly the same thing... have you found any work arounds for this?

**Reply by RobinMalfait** [reply]
You can get the props of any component using `React.ComponentProps<typeof ComponentGoesHere>`, hope that helps!

**Reply by bogdansoare** [reply]
@RobinMalfait doing `React.ComponentProps<typeof ComponentGoesHere>` doesn't work for Dialog

**Reply by giovanniincammicia** [reply]
@RobinMalfait Yep, that's the whole point of the discussion, I tried to use ComponentProps as usual, but I got that error.
[Here is a basic Sandbox](https://codesandbox.io/s/dazzling-joana-gn6rd?file=/src/App.tsx)

**Comment by ernes-to** [comment]
It would be handy if the library also exported the types that each component uses. 

That being said, for anyone else who comes across this issue, you can declare a utility type such as:

```ts
// In a utility types file somewhere in your app
import type { ComponentType } from 'react'

export type ExtractProps<T> = T extends ComponentType<infer P> ? P : T

// Usage in your `.tsx` file
import { Dialog } from '@headlessui/react'

type ModalProps = ExtractProps<typeof Dialog>

export function Modal(props: ModalProps) { ... }
```
Here's a visual example with TypeScript providing intellisense for all available props:

<img src="https://user-images.githubusercontent.com/4595734/149257230-b67694ed-0c8d-4a5b-83e0-9f2cd1e0ff0f.png" width="70%" height="70%" />

Hope it helps 👍

**Reply by will-stone** [reply]
Hi @uicrafts Thanks for this tip. Do you know how to do this with interfaces so it can be extended?

**Comment by ryanprescott** [comment]
> Hi @uicrafts Thanks for this tip. Do you know how to do this with interfaces so it can be extended?

You can extend it with a simple intersection type, but it won't be an interface. For example:

```ts
export type ModalProps = ExtractProps<typeof Dialog> & {
    foo: 'bar' | 'baz',
}
```
